sim/state.ts

```diff
@@ -0,0 +1,340 @@
+					// an object and we need to update the logic in this file to handle it.
+					// The most common case it that someone added a Set/Map which probably
+					// needs to be serialized as an Array/Object respectively - see how
+					// Battle 'hints' or Choice 'switchIns' are handled (and you will likely
+					// need to add the new field to the respective skip constant).
+					throw new TypeError(`Unsupported type ${obj.constructor.name}: ${obj}`);
+				}
+
+				const o: any = {};
+				for (const [key, value] of Object.entries(obj)) {
+					o[key] = this.serializeWithRefs(value, battle);
+				}
+				return o;
+			default:
+				throw new TypeError(`Unexpected typeof === '${typeof obj}': ${obj}`);
+		}
+	}
+
+	private deserializeWithRefs(obj: unknown, battle: Battle) {
+		switch (typeof obj) {
+			case 'undefined':
+			case 'boolean':
+			case 'number':
+				return obj;
+			case 'string':
+				return this.fromRef(obj, battle) || obj;
+			case 'object':
+				if (obj === null) return null;
+				if (Array.isArray(obj)) {
+					const arr = new Array(obj.length);
+					for (const [i, o] of obj.entries()) {
+						arr[i] = this.deserializeWithRefs(o, battle);
+					}
+					return arr;
+				}
+
+				if (this.isActiveMove(obj)) return this.deserializeActiveMove(obj, battle);
+
+				const o: any = {};
+				for (const [key, value] of Object.entries(obj)) {
+					o[key] = this.deserializeWithRefs(value, battle);
+				}
+				return o;
+			case 'function': // lol wtf
+			default:
+				throw new TypeError(`Unexpected typeof === '${typeof obj}': ${obj}`);
+		}
+	}
+
+	private isReferable(obj: object): obj is Referable {
+		// NOTE: see declaration above for why this must be defined lazily.
+		if (!this.REFERABLE) {
+			this.REFERABLE = new Set([
+				Battle, Field, Side, Pokemon, Data.PureEffect,
+				Data.Ability, Data.Item, Data.Move, Data.Template,
+			]);
+		}
+		return this.REFERABLE!.has(obj.constructor);
+	}
+
+	private toRef(obj: Referable): string {
+		// Pokemon's 'id' is not only more verbose than a postion, it also isn't guaranteed
+		// to be uniquely identifying in custom games without Nickname/Species Clause.
+		const id = obj instanceof Pokemon ? `${obj.side.id}${POSITIONS[obj.position]}` : `${obj.id}`;
+		return `[${obj.constructor.name}${id ? ':' : ''}${id}]`;
+	}
+
+	private fromRef(ref: string, battle: Battle): Referable | undefined {
+		// References are sort of fragile - we're mostly just counting on there
+		// being a low chance that some string field in a simulator object will not
+		// 'look' like one. However, it also needs to match one of the Referable
+		// class types to be decode, so we're probably OK. We could make the reference
+		// markers more esoteric with additional sigils etc to avoid collisions, but
+		// we're making a conscious decision to favor readability over robustness.
+		if (ref[0] !== '[' && ref[ref.length - 1] !== ']') return undefined;
```

Comment:
We prefer ``charAt()`` over bracket notation for strings.

Reply:
Also we prefer `ref.slice(-1)` to `ref.charAt(ref.length - 1)`